### ISSUE TITLE
How to resolve CORS issue in v3?

### ISSUE THREAD

**Issue by gezichenshan** [nuxt/framework#6650]
### Environment

Nuxt CLI v3.0.0-rc.6                                                  00:21:35
RootDir: /Users/user/ryanown/ownwork/abroad-go/web                    00:21:38
Nuxt project info:                                                    00:21:38

------------------------------
- Operating System: `Darwin`
- Node Version:     `v16.14.2`
- Nuxt Version:     `3.0.0-rc.6`
- Package Manager:  `npm@8.5.0`
- Builder:          `vite`
- User Config:      `build`, `alias`, `plugins`, `modules`, `css`, `vite`
- Runtime Modules:  `@nuxtjs/tailwindcss@5.3.1`
- Build Modules:    `-`

### Reproduction

Cannot mock a CORS issue here.

### Describe the bug

I got a CORS error when I add a bearer token in my request header:
```
config.headers.Authorization = "Bearer token"
```
Then I got:
`
Access to XMLHttpRequest at 'https://mysite/api/getUsers' from origin 'http://localhost:3000' has been blocked by CORS policy: Request header field authorization is not allowed by Access-Control-Allow-Headers in preflight response.
`
I remember this isn't gonna happen because we can set a proxy server to handle the CORS issue. 

But can can't make it work by setting the nuxt.config.ts lile:
```
vite: {
   server: {
        host: "localhost",
        port: 3000,
        proxy: {
          "/": {
            target: "https://mysite/api",
            changeOrigin: true,
          },
        },
        hmr: { overlay: false },
      },
}
```

The dev server port can even not be changed if I change 3000 to some other numbers like 1997.

haven't it been supported yet?

**Comment by KahrLu** [comment]
Hi @gezichenshan ,I encountered the same problem some time ago. This is my solution. There is official configuration support.
![image](https://user-images.githubusercontent.com/76502123/184764071-f6e875a7-aba0-432e-b696-c7883323250f.png)
related issues no: https://github.com/unjs/nitro/issues/95

**Comment by gezichenshan** [comment]
@KahrLu Thanks for you reply. But where should I add the "routes" to? The doc didn't mention that.

**Comment by KahrLu** [comment]
> @KahrLu Thanks for you reply. But where should I add the "routes" to? The doc didn't mention that.

yes, the document of nuxt3 doesn't mention it. but you can put it in nuxyt.config.ts as like other nuxt config options.
`export default defineNuxtConfig({
    	routes: {
		'/': { prerender: true },
		'/*': { cors: true }
	}
})`

**Comment by gezichenshan** [comment]
@KahrLu  Thanks but I tried that but it still doesn't work. So finally I added "Authorization" in "Access-Control-Allow-Headers" in the backend, then it works.

**Comment by HuakunShen** [comment]
@KahrLu The routes setting only works in development mode for me, but doesn't work in production.

**Comment by HuakunShen** [comment]
@gezichenshan How do you add "Authorization" in "Access-Control-Allow-Headers"?

**Comment by gezichenshan** [comment]
@HuakunShen Like this `c.Header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept, Authorization")`

**Comment by amihhs** [comment]
Maybe this will help us a little. [https://github.com/unjs/nitro/issues/113](https://github.com/unjs/nitro/issues/113)

**Comment by qualiatokyoltd** [comment]
@HuakunShen 

> The routes setting only works in development mode for me, but doesn't work in production.

Too me. may route "cors" options not work.
use server middleware to solution.

at server/middleware/cors.ts

```
export default defineEventHandler((event) => {
  setResponseHeaders(event, {
    "X-Hello-Im-Fine": "thankyou",
    "Access-Control-Allowed-Methods": "GET, POST, PUT, DELETE",
    "Access-Control-Allow-Origin": "*",
  })
})
```

**Comment by chengzhuo5** [comment]
https://github.com/unjs/nitro/issues/539

**Comment by rezaseyedi91** [comment]
@qualiatokyoltd 
> Too me. may route "cors" options not work. use server middleware to solution.
> at server/middleware/cors.ts
> 
> ```
> export default defineEventHandler((event) => {
>   setResponseHeaders(event, {
>     "Access-Control-Allow-Methods": "GET,HEAD,PUT,PATCH,POST,DELETE",
>     "Access-Control-Allow-Origin": "*",
>     'Access-Control-Allow-Credentials': 'true',
>     "Access-Control-Allow-Headers": '*',
>     "Access-Control-Expose-Headers": '*'
>   })
>   if(getMethod(event) === 'OPTIONS'){
>     event.res.statusCode = 204
>     event.res.statusMessage = "No Content."
>     return 'OK'
>   }
> })
> ```

What are 'getMethod'  and 'setResponseHeaders' methods? or where can I import them from?

**Comment by qualiatokyoltd** [comment]
cors props available from nitro ver 0.6.0. no longer have to setting custom header handling.

**Comment by rezaseyedi91** [comment]
> cors props available from nitro ver 0.6.0. no longer have to setting custom header handling.

so far I've tried routeRules confng in nuxt.config.ts>nitro  and nitro.config.ts and these solutions didn't work out. can you describe how and where to use these cors props that you mentioned?
